app/models/user.rb

```diff
@@ -74,4 +74,68 @@ def update_profile(params)
     def refresh_secret
         self.secret = SecureRandom.base64(24)
     end
+
+    def matches
+        all_users = User.all
+        sorted_list = []
+        for other_user in all_users
+            if other_user.id != self.id
+                match_data = self.score_user(other_user)
+                sorted_list.push({:user => other_user, :match_data => match_data})
+            end
+        end
+        return sorted_list.sort { |e1, e2| e1[:match_data][:score] <=> e2[:match_data][:score] } .reverse.last(5)
+    end
+
+    def score_user(other_user)
+        max = 0
+        max_topic = nil
+
+        common_topics = self.find_common_topics(other_user)
+        for topic in common_topics
+            topic_score = self.find_topic_score(other_user, topic)
+            if max < topic_score
+                max = topic_score
+                max_topic = topic
+            end
+        end
+        return {:score => max, :topic => max_topic }
+    end
+
+    def find_common_topics(other_user)
+        my_topics = self.find_topics().to_set
+        other_user_topics = other_user.find_topics().to_set
+        return my_topics & other_user_topics
+    end
+
+    def find_topics()
+        topics = []
+        responses = UserSurveyResponse.where({:user_id => self.id})
+        for response in responses
+            topics.push(response.survey_response.survey_question.topic.id)
+        end
+        return topics
+    end
+
+    def find_topic_score(other_user, topic)
+        sum = 0
+        my_responses_hash = self.build_hash()
+
+        other_responses = UserSurveyResponse.where({:user_id => other_user.id})
+        for other_response in other_responses
+            if other_response.survey_response.survey_question.topic.id == topic
+                sum += ResponseWeight.find_weight(other_response.survey_response.id, my_responses_hash[other_response.survey_response.survey_question.id.to_s])
+            end
+        end
+        return sum
+    end
+
+    def build_hash()
+        my_responses_hash = {}
+        my_responses = UserSurveyResponse.where({:user_id => self.id})
+        for my_response in my_responses
+            my_responses_hash[my_response.survey_response.survey_question.id.to_s] = my_response.survey_response.id
```

Comment:
0.o whoa...... `my_response.survey_response.survey_question.id.to_s` is there a better way to abstract this out? Again, it's fine for now if it's complicated to fix, but we should keep this in mind too.